premium/offline.go

```diff
@@ -41,20 +48,87 @@ var publicKey string
 
 var timeFunc = time.Now
 
-func ValidateLicense(logger zerolog.Logger, meta plugin.Meta, licenseFileOrDirectory string) error {
-	fi, err := os.Stat(licenseFileOrDirectory)
+//go:generate mockgen -package=mocks -destination=../premium/mocks/licensemanager.go -source=offline.go AWSLicenseManagerInterface
+type AWSLicenseManagerInterface interface {
+	CheckoutLicense(ctx context.Context, params *licensemanager.CheckoutLicenseInput, optFns ...func(*licensemanager.Options)) (*licensemanager.CheckoutLicenseOutput, error)
+}
+
+type CQLicenseClient struct {
+	logger                  zerolog.Logger
+	meta                    plugin.Meta
+	licenseFileOrDirectory  string
+	awsLicenseManagerClient AWSLicenseManagerInterface
+}
+
+type LicenseClientOptions func(updater *CQLicenseClient)
+
+func WithMeta(meta plugin.Meta) LicenseClientOptions {
+	return func(cl *CQLicenseClient) {
+		cl.meta = meta
+	}
+}
+
+func WithLicenseFileOrDirectory(licenseFileOrDirectory string) LicenseClientOptions {
+	return func(cl *CQLicenseClient) {
+		cl.licenseFileOrDirectory = licenseFileOrDirectory
+	}
+}
+
+func WithAWSLicenseManagerClient(awsLicenseManagerClient AWSLicenseManagerInterface) LicenseClientOptions {
+	return func(cl *CQLicenseClient) {
+		cl.awsLicenseManagerClient = awsLicenseManagerClient
+	}
+}
+
+func isMarketplaceLicense() bool {
+	return os.Getenv("CQ_AWS_MARKETPLACE_LICENSE") == "true"
+}
+
+func NewLicenseClient(ctx context.Context, logger zerolog.Logger, ops ...LicenseClientOptions) (CQLicenseClient, error) {
+	cl := CQLicenseClient{
+		logger: logger,
+	}
+
+	for _, op := range ops {
+		op(&cl)
+	}
+
+	if isMarketplaceLicense() && cl.awsLicenseManagerClient == nil {
+		cfg, err := awsConfig.LoadDefaultConfig(ctx)
+		if err != nil {
+			return cl, fmt.Errorf("failed to load AWS config: %w", err)
+		}
+		cl.awsLicenseManagerClient = licensemanager.NewFromConfig(cfg)
+	}
+
+	return cl, nil
+}
+
+func (lc CQLicenseClient) ValidateLicense(ctx context.Context) error {
```

Comment:
Also this could be rewritten as a switch (or if elseif) statement and leave the `return ErrLicenseNotApplicable` as the last (or `default` in `switch`) option.